Firestore/Source/Local/FSTLocalDocumentsView.mm

```diff
@@ -111,17 +111,13 @@ - (FSTDocumentDictionary *)documentsMatchingDocumentQuery:(const ResourcePath &)
 
 - (FSTDocumentDictionary *)documentsMatchingCollectionQuery:(FSTQuery *)query {
   // Query the remote documents and overlay mutations.
-  // TODO(mikelehen): There may be significant overlap between the mutations affecting these
-  // remote documents and the allMutationBatchesAffectingQuery mutations. Consider optimizing.
   __block FSTDocumentDictionary *results = [self.remoteDocumentCache documentsMatchingQuery:query];
-  results = [self localDocuments:results];
+  NSArray<FSTMutationBatch *> *matchingBatches =
+      [self.mutationQueue allMutationBatchesAffectingQuery:query];
+  results = [self localDocuments:results inBatches:matchingBatches];
 
-  // Now use the mutation queue to discover any other documents that may match the query after
-  // applying mutations.
   DocumentKeySet matchingKeys;
-  NSArray<FSTMutationBatch *> *matchingMutationBatches =
-      [self.mutationQueue allMutationBatchesAffectingQuery:query];
-  for (FSTMutationBatch *batch in matchingMutationBatches) {
+  for (FSTMutationBatch *batch in matchingBatches) {
```

Comment:
I'm convinced that at the close of this loop, you can assert that matchingKeys will be empty. Try it!

If so, I believe this whole thing can be simplified as to:
  * load the remote documents
  * load the mutation batches
  * apply the mutation batches to results
  * filter results, removing those that don't match

Applying should be like this (sorry for pseudocode):

```
for batch in batches:
  for mutation in batch:
    if mutation.key is not a direct child of query.collection:
      # keys can't change through mutation, so don't bother tracking
      # documents that can't possibly match
      continue

    mutated = [mutation applyTo:results[mutation.key] documentKey:mutation.key]
    apply mutated to results using the logic currently in -localDocuments:
```

Note query filtering doesn't happen on the intermediate states, only on the final.

At that point I think you can delete `localDocuments` and `documentsForKeys`. `localDocument` will have just a single caller and can be inlined into `documentForKey`

Reply:
Unfortunately, four integration tests trigger the case when `matchingKeys` are not empty. They all involve writing new documents to a collection while being offline, e.g. [testGetCollectionWhileOfflineWithDefaultSource](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Integration/API/FIRFirestoreSourceTests.mm#L112-L119). If I understand the code correctly, `remoteCache` only contains the documents which were ack'ed by the server, and `localDocuments` only applies local changes to _given_ documents, which would leave the offline-created `doc4` out of `results` if checking `batches` was omitted.

I'd be happy to get rid of `localDocument` somehow, though. Let me know if you can think of a way around the offline issue.

Reply:
What should be happening is that while those documents aren't in the remote documents cache, they are in the mutation queue, so when you iterate through the batches and apply them to results the new documents should be added.

Reply:
So in addition to [this loop](https://github.com/firebase/firebase-ios-sdk/blob/b26dfc719d484955404d0ad01a7200fbca1ed03f/Firestore/Source/Local/FSTLocalDocumentsView.mm#L186-L197) in `localDocuments`:
```objc
[documents enumerateKeysAndObjectsUsingBlock:^(FSTDocumentKey *key, FSTDocument *remoteDocument,
                                                 BOOL *stop) {
    FSTMaybeDocument *mutatedDoc = [self localDocument:remoteDocument key:key inBatches:batches];
    if ([mutatedDoc isKindOfClass:[FSTDeletedDocument class]]) {
      result = [result dictionaryByRemovingObjectForKey:key];
    } else if ([mutatedDoc isKindOfClass:[FSTDocument class]]) {
      result = [result dictionaryBySettingObject:(FSTDocument *)mutatedDoc forKey:key];
    } else {
      HARD_FAIL("Unknown document: %s", mutatedDoc);
    }
  }];
```
there should be another loop that iterates `batches`, right?

Reply:
Now that I think of it, perhaps there should be just one loop iterating `batches`, and the existing loop can be removed. If a document contains local mutations, it will surely come up while iterating through `batches`.

Reply:
Disregard previous comments, let me try it out...

Reply:
The strategy that localDocuments applies today does not work for what I'm proposing.

It says essentially, for each key already present, apply all mutations for exactly that key to produce the final view of the document.

My proposal in the pseudocode above is to process all mutations that apply to the query such that the mutation will create new documents that don't have a value in remote documents yet.

Note that batches can contain mutations that are in different collections. These batches match the query even though not all the mutations in the batch do. Mutations can't change the key a document though. This is why it's safe to disregard these mutations before the local view has been fully assembled.

Reply:
Does [this](https://gist.github.com/var-const/e5e22399de9147917318dd45ccd9ef4c) look right?

```objc
  __block FSTDocumentDictionary *results = [self.remoteDocumentCache documentsMatchingQuery:query];
  NSArray<FSTMutationBatch *> *matchingBatches =
      [self.mutationQueue allMutationBatchesAffectingQuery:query];

  for (FSTMutationBatch *batch in matchingBatches) {
    for (FSTMutation *mutation in batch.mutations) {
      if (mutation.key.path().PopLast() != query.path) {
        // keys can't change through mutation, so don't bother tracking documents that can't
        // possibly match
        continue;
      }

      FSTDocumentKey* key = static_cast<FSTDocumentKey*>(mutation.key);
      FSTMaybeDocument *baseDoc = results[key];
      FSTMaybeDocument *mutatedDoc =
        [mutation applyTo:baseDoc baseDocument:baseDoc localWriteTime:batch.localWriteTime];

      if ([mutatedDoc isKindOfClass:[FSTDeletedDocument class]]) {
        results = [results dictionaryByRemovingObjectForKey:mutation.key];
      } else if ([mutatedDoc isKindOfClass:[FSTDocument class]]) {
        results = [results dictionaryBySettingObject:(FSTDocument *)mutatedDoc forKey:mutation.key];
      } else {
        HARD_FAIL("Unknown document: %s", mutatedDoc);
      }
    }
  }

  // Note that the extra reference here prevents ARC from deallocating the initial unfiltered
  // results while we're enumerating them.
  FSTDocumentDictionary *unfiltered = results;
  [unfiltered
      enumerateKeysAndObjectsUsingBlock:^(FSTDocumentKey *key, FSTDocument *doc, BOOL *stop) {
        if (![query matchesDocument:doc]) {
          results = [results dictionaryByRemovingObjectForKey:key];
        }
      }];

  return results;
```
(apart from refactoring potential, it actually gives a speed improvement as well)

Reply:
Seems right.

Reply:
Thanks! This function's execution time is now dominated by `dictionaryBySettingObject` (called directly and by remote document cache). Perhaps migrating to C++ version of immutable map might improve timings further.